iree/compiler/Dialect/Flow/Transforms/HLOToHLOPreprocessing.cpp

```diff
@@ -305,6 +305,73 @@ class AdjustDepthwiseFilterShape : public OpRewritePattern<mhlo::ConvOp> {
   }
 };
 
+// clang-format off
+//
+// Rewrites the following pattern
+//
+// %bcastx = "mhlo.broadcast_in_dim"(%x) {broadcast_dimensions = %[[BCAST_DIMS]]} : (%[[SHAPE_BEFORE_BCAST]]) -> %[[SHAPE_AFTER_BCAST]]
+// %bcasty = "mhlo.broadcast_in_dim"(%y) {broadcast_dimensions = %[[BCAST_DIMS]]} : (%[[SHAPE_BEFORE_BCAST]]) -> %[[SHAPE_AFTER_BCAST]]
+// %result = "BinaryElementwiseOpT"(%bcastx, %bcasty) : (%[[SHAPE_AFTER_BCAST]], %[[SHAPE_AFTER_BCAST]]) -> %[[SHAPE_AFTER_BCAST]]
+//
+// into
+//
+// %z = "BinaryElementwiseOpT"(%x, %y) : (%[[SHAPE_BEFORE_BCAST]], %[[SHAPE_BEFORE_BCAST]]) -> %[[SHAPE_BEFORE_BCAST]]
+// %result = "mhlo.broadcast_in_dim"(%z) {} : (%[[SHAPE_BEFORE_BCAST]]) -> %[[SHAPE_AFTER_BCAST]]
+//
+// clang-format on
+template <typename BinaryElementwiseOpT>
+class ReorderBroadcastInDimOpAndBinaryElementwiseOp
+    : public OpRewritePattern<BinaryElementwiseOpT> {
+ public:
+  using OpRewritePattern<BinaryElementwiseOpT>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(BinaryElementwiseOpT op,
+                                PatternRewriter &rewriter) const override {
+    auto lhsType = op.lhs().getType().template dyn_cast<ShapedType>();
+    auto rhsType = op.rhs().getType().template dyn_cast<ShapedType>();
+    if (!lhsType || !rhsType || !lhsType.hasStaticShape() ||
+        lhsType != rhsType) {
+      return failure();
+    }
```

Comment:
In the future we can support this for dynamic shapes where the dynamic shape dimensions have the same SSA values (by relying on CSE to fold any shape math). I think there are still some shape dialect improvements required to make that work, though, so this is more than fine!